std/algorithm/iteration.d

```diff
@@ -4686,6 +4688,613 @@ if (isSomeChar!C)
     }
 }
 
+// substitute
+/**
+Returns a range with all occurrences of `substs` in `r`.
+replaced with their substitution.
+
+Single value replacements (`'ö'.substitute!('ä', 'a', 'ö', 'o', 'ü', 'u)`) are
+supported as well and in $(BIGOH 1).
+
+Params:
+    r = an $(REF_ALTTEXT input range, isInputRange, std,range,primitives)
+    value = a single value which can be substituted in $(BIGOH 1)
+    substs = a set of replacements/substitutions
+
+Returns: a range with the substitutions replaced.
+
+See_Also:
+$(REF replace, std, array) for an eager replace algorithm or
+$(REF translate, std, string), and $(REF tr, std, string)
+for string algorithms with translation tables.
+*/
+template substitute(substs...)
+if (substs.length >= 2 && isExpressions!substs)
+{
+    import std.range.primitives : ElementType;
+    import std.traits : CommonType;
+
+    static assert(!(substs.length & 1), "The number of substitution parameters must be even");
+
+    /**
+      Substitute single values with compile-time substitution mappings.
+      Complexity: $(BIGOH 1) due to D's $(D switch) guaranteeing $(BIGOH 1);
+    */
+    auto substitute(Value)(Value value)
+    if (!isInputRange!Value && !is(CommonType!(Value, typeof(substs[0])) == void))
+    {
+        switch (value)
+        {
+            static foreach (i; 0 .. substs.length / 2)
+                case substs[2 * i]:
+                    return substs[2 * i + 1];
+
+            default: return value;
+        }
+    }
+
+    /// Substitute single range elements with compile-time substitution mappings
+    auto substitute(R)(R r)
+    if (isInputRange!R && !is(CommonType!(ElementType!R, typeof(substs[0])) == void))
+    {
+        return r.map!(a => substitute(a));
+    }
+
+    /// Substitute subranges with compile-time substitution mappings
+    auto substitute(R)(R r)
+    if (isInputRange!R && !is(CommonType!(ElementType!R, ElementType!(typeof(substs[0]))) == void))
+    {
+        // not implemented yet, fallback to runtime variant for now
+        return .substitute(r, substs);
+    }
+}
+
+// In same combinations substitute needs to calculate the auto-decoded length
+// of its needles
+private template needsManualAutodecoding(Range, Needles...)
+{
+    import std.meta : anySatisfy;
+    /* iff
+       - the needles needs auto-decoding, but the incoming range doesn't
+       - (vice versa) if the incoming range needs auto-decoding, but the needle don't
+       - both (range, needle) need auto-decoding and don't share the same common type
+    */
+    enum needlesAreNarrow = anySatisfy!(isNarrowString, Needles);
+    enum sourceIsNarrow = isNarrowString!Range;
+    enum needsManualAutodecoding = ( sourceIsNarrow && !needlesAreNarrow) ||
+                                   (!sourceIsNarrow &&  needlesAreNarrow) ||
+                                   ((sourceIsNarrow && needlesAreNarrow) &&
```

Comment:
no extra parens needed for the three terms in the conjunction

Reply:
I did this for clarity, but ok.